### PR TITLE
Compose reducers

### PR DESCRIPTION
**Pull request by joelevin**
This PRs a lot, but I'm looking for suggestions
As you can see I'm using both case/keyPaths and Optics types, but I would prefer to use case/keyPaths as its more succint, but I'm not sure how to get there without it looking just like the optics types anyways

### CODE REVIEW

File: MotDuJour/ComposableArchitecture/Sources/Reducers.swift [hunk 1]
```diff
@@ -1 +1,11 @@
 public typealias Reducer<Value, Action> = (inout Value, Action) -> Void
+
+public func combine<Value, Action>(_ reducers: (inout Value, Action) -> Void...) -> (
```
Comment (caipre) [inline MotDuJour/ComposableArchitecture/Sources/Reducers.swift]: Nit: can you use the typealias here? `_ reducers: Reducer<Value, Action>...`

File: MotDuJour/Features/Home/Sources/Home.swift [hunk 2]
```diff
@@ -18,3 +18,20 @@ public enum HomeAction {
     case didTapFavorites
     case didTapSearch(query: String)
 }
+
+public enum HomeReducer {
```
Comment (caipre) [inline MotDuJour/Features/Home/Sources/Home.swift]: Nit: the other modules don't have this enum namespace

File: MotDuJour/App/Sources/AppActions.swift [hunk 3]
```diff
@@ -8,12 +11,37 @@ enum AppAction {
...
+            if case let .home(action) = appAction {
+                return action
+            } else {
+                return nil
+            }
+        },
+        embed: { AppAction.home($0) }
+    )
+
+    static let favoritesPrism = Optics.Prism<AppAction, FavoritesAction>(
```
Comment (caipre) [inline MotDuJour/App/Sources/AppActions.swift:27]: Is this different from the CasePath for `AppAction.favorites`?

**Comment by caipre** [comment]
> I'm not sure how to get there without it looking just like the optics types anyways

I think structural typing (aka duck typing) would be the way to go, but it's not something that exists outside of TypeScript and a few other languages, so far as I know. The idea being that you don't strictly need a `FooState` struct, but just something that conforms to the same shape. It's looser than static typing, but still guarantees that the type has the properties you expect. The problem of course being that it the data might not be "correct" even though it has the same shape.

This is what they do in the videos, I think. An alternative we could try is to just embed the various structs into AppState. That's not really buying us much, though, since each state would be totally separate then, with all the problems of consistency that that would imply.

File: MotDuJour/ComposableArchitecture/Sources/Reducers.swift (hunk 1 above)
Comment (joelevin) [reply MotDuJour/ComposableArchitecture/Sources/Reducers.swift]: yeah for sure

File: MotDuJour/Features/Home/Sources/Home.swift (hunk 2 above)
Comment (joelevin) [reply MotDuJour/Features/Home/Sources/Home.swift]: yep, is this namespace something we want? I'm on the fence

File: MotDuJour/App/Sources/AppActions.swift (hunk 3 above)
Comment (joelevin) [reply MotDuJour/App/Sources/AppActions.swift:27]: no, but we don't have an API that supports the prism being a CasePath and the lens being a `Optics.Lens`; either the lens and prisms are both case/keypaths or both `Optics`

**Comment by joelevin** [comment]
> > I'm not sure how to get there without it looking just like the optics types anyways
> 
> I think structural typing (aka duck typing) would be the way to go, but it's not something that exists outside of TypeScript and a few other languages, so far as I know. The idea being that you don't strictly need a `FooState` struct, but just something that conforms to the same shape. It's looser than static typing, but still guarantees that the type has the properties you expect. The problem of course being that it the data might not be "correct" even though it has the same shape.
> 
> This is what they do in the videos, I think. An alternative we could try is to just embed the various structs into AppState. That's not really buying us much, though, since each state would be totally separate then, with all the problems of consistency that that would imply.

Something like...
`struct AppState {`
`let homeState: HomeState`
`let favoritesState: FavoritesState`
`}`
?

I agree, seems like it's missing the point

**Comment by joelevin** [comment]
@caipre I watched the next video and added the computed vars extension as an alternative, now this PR has both compositions of reducers so we should choose one that we like the look of more

**Review by caipre — APPROVED** [review]

**Pull request merged**